crates/router/src/core/payments/helpers.rs

```diff
@@ -2060,27 +2060,47 @@ pub fn generate_mandate(
 }
 
 // A function to manually authenticate the client secret with intent fulfillment time
-pub fn authenticate_client_secret(
+pub async fn authenticate_client_secret(
     request_client_secret: Option<&String>,
     payment_intent: &PaymentIntent,
     merchant_intent_fulfillment_time: Option<i64>,
+    db: &dyn StorageInterface,
 ) -> Result<(), errors::ApiErrorResponse> {
+    let payment_link_db = if let Some(payment_link_id) = &payment_intent.payment_link_id {
```

Comment:
Why not move this inside the client_secret verification check? It may take a db hit even with a wrong client secret?